blueprints/user.permissions.md

```diff
@@ -0,0 +1,286 @@
+	phoneNumber: string | null;
+	postalCode: string | null;
+	publicSSHKey: string | null;
+	registrationSent: string | null;
+	role: string;
+	stateOrProvince: string | null;
+	tenant?: string;
+	tenantID: number;
+	uid: number | null;
+	username: string;
+}
+```
+
+... replacing the required `role` field that was a Role ID with one that is
+the Role's name (and therefore removing `roleName` as it's now redundant).
+
+(Note that `lastUpdated` changed from `string | null` to `Date`, which
+represents the transition to using proper RFC3339 formatting, and also some
+fields which cannot be `null` for a valid user are no longer `| null`. Those
+aren't changes necessary to implement Permissions, but can be done easily at
+the same time as necessary changes to User.)
+
+## Component Impact
+This change primarily concerns Traffic Ops and Traffic Portal, though any
+client of the API will need to be aware of the changes to its authorization
+system.
+
+### Traffic Portal Impact
+There is a "Capabilities" view currently in Traffic Portal that provides a
+table of user Permissions, though it is not linked to the side navigation bar.
+This should be removed.
+
+The Roles' details pages will need to be augmented with controls for adding and
+removing Permissions. The Permissions available for adding to a user should
+include an autocompletion of those known by Traffic Portal to exist, but
+shouldn't restrict the addition of Permissions not known to exist to accomodate
+running older versions of Traffic Portal with newer versions of Traffic Ops, as
+well as any plugins unbeknownst to Traffic Portal that add Traffic Ops API
+endpoints and may declare their own Permissions that do not exist in a vanilla
+Traffic Control environment.
+
+Also, the Roles table will be augmented with the ability to filter Roles by the
+presence or absence of a Permission, although the full itemization of the
+Permissions afforded to a Role would, in most cases, be far too large to
+comfortably display in the table itself.
+
+Under-the-hood, API services will need to change from using Role IDs to Role
+names.
+
+### Traffic Ops Impact
+Traffic Ops will reqire changes to its Roles and Permissions API, chiefly, and
+the database tables that back them. Enforcement of API Permissions, though,
+will also have more wide-ranging impact that will require changes to very
+nearly every API endpoint.
+
+Additionally, a configuation file field will be added named `usePermissions`
+that is an optional boolean which, when present and `true` causes Traffic Ops
+to use Permissions rather than Role Privilege Level to determine a user's
+authorization for a given operation.
+
+#### API Impact
+Structurally, the only necessary changes are to the `/roles` and `/users`
+endpoints, which will need to be updated to output structures consistent with
+the changes outlined in the Data Model Impact section. The `/capabilities` and
+`/api_capabilities` endpoints will be removed (and we might consider renaming
```

Comment:
> and we might consider renaming the `/server_capabilities` and `/server_server_capabilities` to drop the now-unnecessary "server_" qualifier

seems beyond the scope of this change. if you agree, any reason to mention it?

Reply:
Because the only reason those have that qualifier is to distinguish them from Role Capabilities, which will no longer exist.

Reply:
I don't think they were named that way just to distinguish them from Role Capabilities. They were specifically called Server Capabilities because they are the capabilities of a server. Renaming the endpoints would seem like an avoidable breaking change anyways.

Reply:
I definitely remember the confusion between Capabilities and Server Capabilities being discussed explicitly when deciding on the names of API routes - but that doesn't really matter. I just thought it was worth calling out, but it's a waste of everyone's time to have this point of contention for something that really is pretty unrelated to what the blueprint's actually about. I'll just remove it.